Implement linear module for APoT quantization

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82105

### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

asl3 committed:
Implement linear module for APoT quantization

[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"




[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"




[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"




[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"




[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"


### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,180 @@
+    """
+
+    def __init__(self, weight2quantize: torch.Tensor, signed=False):
+        # self.in_features = in_features
```
should these be removed?

HDCharles commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,180 @@
+        return self.linear_APoT_fn(activation)
+
+
+    def from_reference(self,  # type: ignore[override]
```
shouldn't this be a class method?

HDCharles commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,180 @@
+
+    def decompose_APoT(self):
+        r"""
+        Helper function to decompose sums of PoT from APoT quantization
```
it seems like this significantly recreates the Apot datatype, can you give some information on the output i.e. something like levels_decomposed[i][j] is the ith weight of the jth power of two or something along those lines

HDCharles commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,180 @@
+        p_all = []
+
+        # create the "levels"
+        p0 = torch.tensor([0, 2**0, 2**-2, 2**-4])
```
its weird that this is hardcoded, makes a lot more sense to set B, K as object vars and use them throughout the derive such values

asl3 replied:
since i need to decompose the weight values it was easier to write it out explicitly like this -- since b and k are hardcoded (like how scale and zero_point are hardcoded in [uniform linear](https://www.internalfb.com/code/fbsource/[adba684270a7242e9856163a61480777f88fd57f]/fbcode/caffe2/torch/nn/quantized/modules/linear.py?lines=237) i thought this should be ok

HDCharles commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,180 @@
+
+        return result
+
+    def forward(self, activation: torch.Tensor) -> torch.Tensor:
```
whats the output dtype supposed to be?

asl3 replied:
fp32 (added specification in comments)

HDCharles commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,180 @@
+        Multiply APoT quantized weight and uniformly quantized activation
+        with bitshifting instead of matrix multiplication.
+        Args:
+            activation (Tensor): uniformly quantized activation tensor
```
can you specify the dtype of the activation, a uniformly quantized activation tensor would have dtype quint8, whereas the test takes in a uint8, it should also be noted that simply bit shifting the integer in a quint8 will not give the correct output when z!=0 but if youo also bitshift z then it should work, though the recombination will be problematic.

In general i'm not sure if this makes sense to combine with uniform quantization since its not a fixed point representation.

HDCharles commented on test/quantization/core/experimental/test_linear.py:
```diff
@@ -0,0 +1,40 @@
+    """
+        Test linear_APoT_fn by calling forward method
+    """
+    def test_linear_APoT_fn(self):
```
you should compare with an actual linear. This basically redoes the bit shift implementation as a multiply but there's more going on than just that which needs to be tested. A full end to end comparison with a float linear would help with that.

asl3 committed:
Update on "Implement linear module for APoT quantization"


### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"


### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

[ghstack-poisoned]

asl3 committed:
Update on "Implement linear module for APoT quantization"


### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

[ghstack-poisoned]

dzdang commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,158 @@
+
+        return blocks
+
+    def bitshift_multiplication(self, decomposed_weight, activation):
```
the naming for this function should reflect what's done in the functino. in this function, you're doing matmul, not just bitshift multiplication. you could have a matmul function and refactor that bitshift into another method called `bitshift_mul...` and call it from within matmul.

asl3 replied:
fixed!

dzdang reviewed (commented):
will finish the review later

HDCharles commented on test/quantization/core/experimental/test_linear.py:
```diff
@@ -0,0 +1,37 @@
+        # weight: fp tensor
+        weight = 1000 * torch.rand(4, 4)
+
+        # activtion: uniformly quantized tensor
```
nit: this isn't a quantized tensor, its an fp32 tensor with ~ integer values

asl3 committed:
Update on "Implement linear module for APoT quantization"


### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

[ghstack-poisoned]

dzdang commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+
+        return blocks
+
+    def bitshift_mul(self, decomposed_weight, activation):
```
you're still doing the matrix multiplication in this function. it doesn't address my previous comment. what you have here isn't just doing bitshift_mul. it's doing the entire matrix multiplication

dzdang commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+        r"""
+        Decompose APoT quantized terms into binary digits
+        Args:
+            x (Tensor): binary representation of APoT quantized tensor
```
is this comment correct?

asl3 replied:
changed to "Decompose binary representation of APoT values into list of k-sized blocks"

dzdang commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+
+        self.alpha, self.gamma, self.quantization_levels, self.level_indices = observer.calculate_qparams(signed=False)
+
+        quantized = quantize_APoT(weight2quantize, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
```
quantized_weight

dzdang commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+        self.weight_transposed = torch.transpose(self.weight, 0, 1)
+
+        self.weight_transposed = self.weight_transposed.reshape(weight2quantize.shape)
+        self.weight_transposed = self.weight_transposed.reshape(weight2quantize.shape[1], weight2quantize.shape[0])
```
why are these reshapes necessary?

asl3 replied:
yeah not necessary, removed

dzdang commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+        super().__init__()
+
+        # hard code b, k to match uniform quantization: b=4, k=1
+        self.b = 8
```
i'm a bit confused. why are we hardcoding these instead of having them as default parameters? your bitshift matrix multiplication should be generic and shouldn't be limited to b = 8 and k = 1...

asl3 replied:
the uniform linear module scale hardcodes scale=1 and zero_point=0 so i was following that, but i can take b and k as input

dzdang reviewed (changes requested):
it seems your implementation is hard coded for b=8, k = 1. it should apply generically for any valid combination of b and k

jerryzh168 commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+                                  produced by torch.ao.quantization.experimental
+                                  utilities or provided by the user
+        """
+        qlinear = ref_qlinear
```
I think we'll need to create a `LinearAPoT` instance here, for example: https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/linear.py#L290

jerryzh168 replied:
also this path is not yet tested, maybe we can leave this unimplemented for now

asl3 replied:
sounds good, will leave unimplemented

jerryzh168 commented on torch/ao/quantization/experimental/linear.py:
```diff
@@ -0,0 +1,168 @@
+        Args:
+            activation (Tensor): uniformly quantized activation tensor
+        """
+        return self.linear_APoT_fn(activation)
```
nit: looks like linear_APoT_fn is only used by this function, why not just inline the code in linear_APoT_fn here?

asl3 committed:
Update on "Implement linear module for APoT quantization"


### Summary
Implement linear module to support APoT quantization. Use bitshifting method discussed in APoT paper https://arxiv.org/pdf/1909.13144.pdf to multiply PoT terms in APoT quantized weight tensor with uniformly quantized activation tensor to demonstrate alternative to matrix multiplication.

Multiplication using bitshifting for PoT: 

<img width="340" alt="Screen Shot 2022-07-25 at 12 44 26 PM" src="https://user-images.githubusercontent.com/68875504/180831050-ff849bca-8eb0-4b69-9b7f-c6c94a4cdfb5.png">

### Test Plan
Run unit tests with: `python /pytorch/test/quantization/core/experimental/test_linear.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
